packages/firestore/src/local/shared_client_state.ts

```diff
@@ -376,18 +495,28 @@ export class WebStorageSharedClientState implements SharedClientState {
         event.key !== this.localClientStorageKey,
         'Received LocalStorage notification for local change.'
       );
-      const clientKey = this.fromLocalStorageClientKey(event.key);
-      if (clientKey) {
-        if (event.newValue == null) {
-          delete this.activeClients[clientKey];
-        } else {
-          const newClient = RemoteClientState.fromLocalStorageEntry(
-            clientKey,
-            event.newValue
-          );
-          if (newClient) {
-            this.activeClients[newClient.clientKey] = newClient;
-          }
+
+      if (event.newValue == null) {
+        const clientId = this.fromLocalStorageClientStateKey(event.key);
+        if (clientId != null) {
+          delete this.activeClients[clientId];
+        }
+      } else {
+        const clientState = this.fromLocalStorageClientState(
+          event.key,
+          event.newValue
+        );
+        if (clientState) {
+          this.activeClients[clientState.clientId] = clientState;
+          return;
+        }
+        const mutationBatch = this.fromLocalStorageMutation(
+          event.key,
+          event.newValue
+        );
+        if (mutationBatch) {
+          this.handleMutationBatchEvent(mutationBatch);
+          return;
```

Comment:
I'm not a big fan of this code structure, calling fromBlah() indiscriminately and using nulls to decide whether the thing you passed to fromBlah() was actually a Blah, and using early-returns...  Bunch of minor code smells that I don't enjoy. :-)

I'd be much happier doing a bit of extra work (e.g. running the regex twice) but making the logic more explicit...  something like:

```
if (this.clientStateKeyRe.test(event.key)) {
  ...
} else if (this.mutationBatchKeyRe.test(event.key)) {
  ...
}
```